types/decimal.go

```diff
@@ -0,0 +1,324 @@
+	return multiplier
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDec(i, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(big.NewInt(i), precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromBigInt(i *big.Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i, precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromInt(i Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i.BigInt(), precisionMultiplier(prec)),
+	}
+}
+
+// create a decimal from a decimal string (ex. "1234.5678")
+func NewDecFromStr(str string) (d Dec, err Error) {
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	// first extract any negative symbol
+	neg := false
+	if string(str[0]) == "-" {
+		neg = true
+		str = str[1:]
+	}
+
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	strs := strings.Split(str, ".")
+	lenDecs := 0
+	combinedStr := strs[0]
+	if len(strs) == 2 {
+		lenDecs = len(strs[1])
+		if lenDecs == 0 || len(combinedStr) == 0 {
+			return d, ErrUnknownRequest("bad decimal length")
+		}
+		combinedStr = combinedStr + strs[1]
+	} else if len(strs) > 2 {
+		return d, ErrUnknownRequest("too many periods to be a decimal string")
+	}
+
+	if lenDecs > Precision {
+		return d, ErrUnknownRequest("too much Precision in decimal")
+	}
+
+	// add some extra zero's to correct to the Precision factor
+	zerosToAdd := Precision - lenDecs
+	zeros := fmt.Sprintf(`%0`+strconv.Itoa(zerosToAdd)+`s`, "")
+	combinedStr = combinedStr + zeros
+
+	combined, ok := new(big.Int).SetString(combinedStr, 10)
+	if !ok {
+		return d, ErrUnknownRequest("bad string to integer conversion")
+	}
+	if neg {
+		combined = new(big.Int).Neg(combined)
+	}
+	return Dec{combined}, nil
+}
+
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return !d.LT(d2) }                            // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }            // less than
+func (d Dec) LTE(d2 Dec) bool   { return !d.GT(d2) }                            // less than or equal
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} }         // Is equal to zero
+func (d Dec) Add(d2 Dec) Dec    { return Dec{new(big.Int).Add(d.Int, d2.Int)} } // addition
+func (d Dec) Sub(d2 Dec) Dec    { return Dec{new(big.Int).Sub(d.Int, d2.Int)} } // subtraction
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := BankerRoundChop(mul, Precision)
```

Comment:
Lets check for overflow here, like we do with normal Sdk Ints. (We don't want it to be over 256 bits + prec_bits long)

Reply:
How do we calculate the bits used for 10 decimals of precision? haven't resolved yet (feel free to push fyi over the weekend - but I'll get to it on tuesday otherwise)

Reply:
If we want a fast heuristic, we can cache the base2 log of `precision`, and set `prec_bits = ceil(log_2(precision))`. (We could then do this heuristic before computation as well),

Reply:
cool - yeah the 33.2 makes sense to me, (appreciate the explanation!), I think the simplest thing is to perform the computation on the resulting value. doing this we can just use the value of `chopped` which already has 10 decimal places thus the check now becomes:

```golang
	if chopped.BitLen() > 289 {
		panic("Int overflow")
	}
```

This way we can use the same overflow check for all computations without having to consider how to determine the bytes required for each operation type based on the length of the input variables. (as in you're adding the bytes for a multiplication, what about for the rest of BEDMAS family?)

Reply:
addressed added in these sanity checks for Add/Mul/Quo/Sub as well as RoundInt64